sdk/core/azure-core/src/main/java/com/azure/core/util/Option.java

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import java.util.NoSuchElementException;
+
+/**
+ * The Option type to describe tri-state. Every Option instance is in one of
+ * the three states: a state representing a non-null-value, null-value, or no-value.
+ *
+ * @param <T> The value type.
+ */
+public final class Option<T> {
+    private static final Option<?> UNINITIALIZED = new Option<>();
+    private static final Option<?> EMPTY = new Option<>(null);
+    private final boolean isInitialized;
+    private final T value;
+
+    /**
+     * Returns an {@link Option} with the specified null-value or non-null-value.
+     *
+     * @param <T> The value type.
+     * @param value the value.
+     * @return an {@link Option} with the value present.
+     */
+    public static <T> Option<T> of(T value) {
+        return value == null ? empty() : new Option<>(value);
+    }
+
+    /**
+     * Returns an {@link Option} with null-value.
+     * <p>
+     * {@code Option.empty()} is a syntactic sugar for {@code Option.of(null)}.
+     * </p>
+     * @param <T> The value type.
+     * @return an {@link Option} with a null-value.
+     */
+    public static <T> Option<T> empty() {
+        @SuppressWarnings("unchecked")
+        Option<T> empty = (Option<T>) EMPTY;
+        return empty;
+    }
+
+    /**
+     * Returns an {@link Option} instance with no-value.
+     *
+     * @param <T> Type of the non-existent value.
+     * @return An Option type with no-value.
+     */
+    public static <T> Option<T> uninitialized() {
+        @SuppressWarnings("unchecked")
+        Option<T> uninitialized = (Option<T>) UNINITIALIZED;
+        return uninitialized;
+    }
+
+    /**
+     * Return {@code true} if this instance is initialized with a null-value or non-null-value,
+     * otherwise {@code false}.
+     *
+     * @return {@code true} if there is a value present, otherwise {@code false}
+     */
+    public boolean isInitialized() {
+        return this.isInitialized;
+    }
+
+    /**
+     * Gets the value in the {@link Option}.
+     *
+     * @return The value.
+     * @throws NoSuchElementException thrown if the {@link Option} has no value.
+     */
+    public T getValue() {
```

Comment:
We need a `setValue()` method as well. Otherwise, this will be immutable resulting in many `Option` objects being created to mutate the value.

Reply:
was thinking about it, the current "known" use case is - This type will be used for JSON-merge-patch input only model. 

The JSON-merge-patch model type will not have any required property constraints. If the user wants to clear a property in the server they set it to "Option.of(null)" and to set a value "Option.of(non-null)". The instantiation of such a model and setting of its properties highly likely to happen in one place along with an update API call.

In this known use case, I couldn't find a need to make this type mutable, unless we foresee the immediate application of this type in any other use case. So I was preferring to keep it simple & immutable, like JDK's Optional type.

Please let me know your thoughts.

Reply:
Yeah, I am fine with just keeping this immutable for now and we can extend it later if there's a usecase for it.